packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+    Partial<ITransactionWithSignatures>;
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function partiallySignTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & ITransactionWithSignatures> {
+    const { partialSigners, modifyingSigners } = categorizeTransactionSigners(
+        deduplicateSigners([...getSignersFromTransaction(transaction).filter(isTransactionSigner), ...signers]),
+        false
+    );
+
+    return signModifyingAndPartialTransactionSigners(transaction, modifyingSigners, partialSigners);
+}
+
+/**
+ * Signs a transaction using any signers that may be stored in IAccountSignerMeta instruction accounts
+ * as well as any signers provided explicitly to this function.
+ * It will assert that the transaction is fully signed before returning.
+ * It will ignore TransactionSendingSigners since this function does not send the transaction.
+ */
+export async function signTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = []
+): Promise<TTransaction & IFullySignedTransaction> {
+    const signedTransaction = await partiallySignTransactionWithSigners(transaction, signers);
+    assertTransactionIsFullySigned(signedTransaction);
+    return signedTransaction;
+}
+
+/**
+ * Signs and sends a transaction using any signers that may be stored in IAccountSignerMeta
+ * instruction accounts as well as any signers provided explicitly to this function.
+ * It will identify a single TransactionSendingSigners to use for sending the transaction, if any.
+ * Otherwise, it will send the transaction using the provided fallbackSender.
+ */
+export async function signAndSendTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = [],
+    fallbackSender?: (transaction: TTransaction & IFullySignedTransaction) => Promise<SignatureBytes>
+): Promise<SignatureBytes> {
+    const { partialSigners, modifyingSigners, sendingSigner } = categorizeTransactionSigners(
+        deduplicateSigners([...getSignersFromTransaction(transaction).filter(isTransactionSigner), ...signers])
+    );
+
+    const signedTransaction = await signModifyingAndPartialTransactionSigners(
+        transaction,
+        modifyingSigners,
+        partialSigners
+    );
+
+    if (sendingSigner) {
+        const [signature] = await sendingSigner.signAndSendTransactions([signedTransaction]);
```

Comment:
100% agree. I was talking to Jordan about this yesterday and I think we should have a second optional `configs` argument that enables users to provide their abort signals.

I'll make a PR for it asap.

Reply:
**PR available here: https://github.com/solana-labs/solana-web3.js/pull/1847**

I found it more convenient to append the PR (instead of prepending it) so we can see the full abort signal changes in one go. I'll make sure both PRs are merged together when they are approved.